lstm/bmi_lstm.py

```diff
@@ -1,1046 +1,562 @@
+
+        # load static variables from config into state
+        load_static_attributes(self.cfg_bmi, self._static_inputs)
+
+    def update(self) -> None:
+        """update a single timestep."""
+
+        # wrap dynamic and static inputs in a container so an `EnsembleMember`
+        # can access both like they are from the same `State` object.
+        #
+        # each ensemble member will query the `state` object for its required inputs.
+        # this could ensemble members with a different number of required features in the future.
+        state = StateFacade(self._dynamic_inputs, self._static_inputs)
+
+        outputs: dict[str, list[float]] = collections.defaultdict(list)
+        for member in self.ensemble_members:
+            for output in member.update(state):
+                assert len(output.value) == 1, (
+                    f"expected output of length 1, got {len(output.value)}"
+                )
+                outputs[output.name].append(output.value[0])
+
+        # ensemble output and set output variables
+        for name, values in outputs.items():
+            self._outputs.set_value(name, np.mean(values, dtype="float64"))
+
+        # increment model timestep
+        self._timestep += 1
+
+    def update_until(self, time: float) -> None:
+        if time <= self.get_current_time():
+            current_time = self.get_current_time()
+            logger.warning(f"no update performed: {time=} <= {current_time=}")
+            return None
+
+        n_steps, remainder = divmod(
+            time - self.get_current_time(), self.get_time_step()
+        )
+
+        if remainder != 0:
+            logger.warning(
+                f"time is not multiple of time step size. updating until: {time - remainder=} "
+            )
 
-        Parameters
-        ----------
-        item
-            The name of the attribute item to get.
+        for _ in range(int(n_steps)):
+            self.update()
 
-        Returns
-        -------
-        The value of the named item.
-        """
-        # Have these work explicitly (or else loops)
-        if item == '_input_var_names' or item == '_output_var_names':
-            return super(bmi_LSTM, self).__getattribute__(item)
-
-        # By default, for things other than BMI variables, use normal behavior
-        if item not in super(bmi_LSTM, self).__getattribute__('_input_var_names') and item not in super(bmi_LSTM, self).__getattribute__('_output_var_names'):
-            return super(bmi_LSTM, self).__getattribute__(item)
-
-        # Return the single scalar value from any ndarray of size 1
-        value = super(bmi_LSTM, self).__getattribute__(item)
-        if isinstance(value, np.ndarray) and value.size == 1:
-            return value[0]
-        else:
-            return value
-
-    def __setattr__(self, key, value):
-        """
-        Customized instance attribute mutator functionality.
+    def finalize(self) -> None: ...
```

Comment:
It may be worth explicitly `del` ing the esemble memebers if the underlying torch mechanics hold a lot of resource that we don't want hanging around once something has called this function to indicate its no longer needed (but perhaps is still referencing the BMI handle)

Reply:
Yeah that is fair. I think we should keep it as simple as possible until it shows up as a problem.

Reply:
This can go into an issue for a future enhancement, for sure.